#MIT 6.036 Spring 2019: Homework 12

In [2]:
import numpy as np
import time
import pickle

#Setup
Run the next code block to download and import the code for this lab.



In [21]:
!wget --quiet https://introml_oll.odl.mit.edu/6.036/static/homework/hw12/data_for_hw12.zip
!unzip data_for_hw12.zip

Archive:  data_for_hw12.zip
  inflating: movies.csv              
  inflating: ratings.csv             


#Some preliminary code

Here are some useful functions which will be used in the following sections:

In [3]:
def pred(data, x):
    (a, i, r) = data
    (u, b_u, v, b_v) = x
    return np.dot(u[a].T,v[i]) + b_u[a] + b_v[i]

# X : n x k
# Y : n
def ridge_analytic(X, Y, lam):
    (n, k) = X.shape
    xm = np.mean(X, axis = 0, keepdims = True)   # 1 x n
    ym = np.mean(Y)                              # 1 x 1
    Z = X - xm                                   # d x n
    T = Y - ym                                   # 1 x n
    th = np.linalg.solve(np.dot(Z.T, Z) + lam * np.identity(k), np.dot(Z.T, T))
    # th_0 account for the centering
    th_0 = (ym - np.dot(xm, th))                 # 1 x 1
    return th.reshape((k,1)), float(th_0)

## 2) Some movies are more equal than others

In Lab 12, we formulated recommender systems without offsets:  there was no equivalent of $\theta_0$ in the regression problem we solved.  But offsets are very useful in this problem to account for some reviewers that are just generally grumpy (or enthusiastic) or some movies that are just generally awful (or great). For this section we will work to extend the results from the lab to include offsets.

Rather than trying to predict what rating Amy will give a movie based on the properties of that movie, it might be more effective to predict how much "more highly than usual" Amy will rate the movie.  To do this, we introduce a vector $b_u$ to characterize offsets for users and a vector $b_v$ to characterize offsets for movies.  Now, the objective can be written as
$$J(U, V) = \frac{1}{2}\sum_{(a, i) \in D} (Y_{ai} - {u^{(a)}}\cdot v^{(i)} - b_u^{(a)} - b_v^{(i)})^2 + \frac{\lambda}{2} \sum_{a = 1}^n \lVert u^{(a)} \rVert^2 + \frac{\lambda}{2} \sum_{i = 1}^m  \lVert v^{(i)} \rVert^2 $$

We will stick with just one half of the problem for now:  finding $U$ and $b_u$ given a fixed $V$ and $b_v$.  We'll concentrate on computing a new estimate for $u^{(a)}$ and $b_u^{(a)}$.  Continuing the notation from lab, the regression problem that we have to solve becomes:
$$-B_a^T (Z_a - B_a u^{(a)} - b_v - b_u^{(a)}) + \lambda u^{(a)} = 0$$
where $b_v$ is the $l_a$ by 1 vector of offsets for the movies and $b_u^{(a)}$ is the offset for user $a$.  This is a linear regression problem in which the targets can be viewed as $Z_a - b_v$, and where there is an offset $b_u^{(a)}$ that is not regularized.  This is, then, the standard ridge regression set-up.  We have seen how to solve this problem via gradient descent before, but it can also be solved analytically by essentially turning it into a regular linear regression problem. Using this approach to computing the optimal $u^{(a)}$ requires first centering the data, then doing linear regression, then doing a little bit of work to recover the offset $b_u^{(a)}$.  In the code, we have supplied a procedure (`ridge_analytic`) for doing exactly this.

Let's see how it works out in our running example.  For illustrative purposes, we'll add two more movies:

8) 6.036 lecture videos <br />
9) The Zzzz files <br />

Our new preference data is:
$$Y = \begin{bmatrix}
? & 1 & ? & 1 & 5 & 1 & 5 & 5 & 1\\
1 & 5 & 1 & ? & ? & 5 & 1 & 5 & 1\\
5 & 5 & 5 & 5 & ? & ? & 5 & ? & 1\\
1 & ? & 1 & 1 & 1 & 1 & ? & 5 & ?
\end{bmatrix}$$

We will assume the following value for $b_v$:
$$b_v = \begin{bmatrix}
3 & 3 & 3 & 3 & 3 & 3 & 3 & 5 & 1
\end{bmatrix}^T $$
And let
$$V = \begin{bmatrix}
10 & 1 & 10 & 1 & 10 & 1 & 10 & 5 & 5\\
1 & 10 & 1 & 10 & 1 & 10 & 1 & 5 & 5
\end{bmatrix}
$$
We can now compute the optimal $u^{(a)}$ and $b_u^{(a)}$ for $a=0$:

In [4]:
Z = np.array([[1], [1], [5], [1], [5], [5], [1]])
b_v = np.array([[3], [3], [3], [3], [3], [5], [1]])
B = np.array([[1, 10], [1, 10], [10, 1], [1, 10], [10, 1], [5, 5], [5, 5]])
# Solution with offsets, using ridge_analytic provided in code file
u_a, b_u_a = ridge_analytic(B, (Z - b_v), 1)

#u_a, b_u_a
#(array([[ 0.22024566],
#        [-0.22193986]]),
# array([[ 0.00762389]]))

# Solution using previous model, with no offsets
u_a_no_b = np.dot(np.linalg.inv(np.dot(B.T, B) + 1 * np.identity(2)), np.dot(B.T, Z))

#u_a_no_b
#array([[ 0.50148126],
#       [ 0.0562376 ]])

<ipython-input-3-a3ba3405c5ad>:17: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return th.reshape((k,1)), float(th_0)


Based on this, we can predict Amy's ratings on new Movies!

1) How will Amy feel about a brand new Llama movie (not in the existing movie data) that gets bad ratings from almost everyone ($b_v^{(i)}= 1$), in the two models?

In [6]:
#Code for answering (1) here
import numpy as np

# Given data
Z = np.array([[1], [1], [5], [1], [5], [5], [1]])
b_v = np.array([[3], [3], [3], [3], [3], [5], [1]])
B = np.array([[1, 10], [1, 10], [10, 1], [1, 10], [10, 1], [5, 5], [5, 5]])

# Ridge regression function
def ridge_analytic(X, Y, lam):
    (n, k) = X.shape
    xm = np.mean(X, axis=0, keepdims=True)   # 1 x k
    ym = np.mean(Y)                          # scalar
    Z = X - xm                               # n x k
    T = Y - ym                               # n x 1
    th = np.linalg.solve(np.dot(Z.T, Z) + lam * np.identity(k), np.dot(Z.T, T))
    th_0 = (ym - np.dot(xm, th))             # scalar
    return th.reshape((k, 1)), float(th_0)

# Compute u_a and b_u_a using ridge_analytic
u_a, b_u_a = ridge_analytic(B, (Z - b_v), 1)

# Solution without offsets
u_a_no_b = np.dot(np.linalg.inv(np.dot(B.T, B) + 1 * np.identity(2)), np.dot(B.T, Z))

# Feature vector for the Llama movie
v_llama = np.array([[10], [1]])

# Prediction without offset
prediction_no_offset = np.dot(u_a_no_b.T, v_llama) + 1.0

# Prediction with offset
prediction_with_offset = np.dot(u_a.T, v_llama) + b_u_a

# Prepare the result
result = [float(f"{prediction_no_offset[0][0]:.2f}"), float(f"{prediction_with_offset[0][0]:.2f}")]
print(result)  # Output the predictions


[6.07, 1.99]


<ipython-input-6-3727fd5b236d>:18: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return th.reshape((k, 1)), float(th_0)


2) What about a brand new Robot movie (not in the existing movie data) that gets good ratings from almost everyone ($b_v^{(i)}= 5$), in the two models?


In [7]:
#Code for answering (2) here
import numpy as np

# Given data
Z = np.array([[1], [1], [5], [1], [5], [5], [1]])
b_v = np.array([[3], [3], [3], [3], [3], [5], [1]])
B = np.array([[1, 10], [1, 10], [10, 1], [1, 10], [10, 1], [5, 5], [5, 5]])

# Ridge regression function
def ridge_analytic(X, Y, lam):
    (n, k) = X.shape
    xm = np.mean(X, axis=0, keepdims=True)   # 1 x k
    ym = np.mean(Y)                          # scalar
    Z = X - xm                               # n x k
    T = Y - ym                               # n x 1
    th = np.linalg.solve(np.dot(Z.T, Z) + lam * np.identity(k), np.dot(Z.T, T))
    th_0 = (ym - np.dot(xm, th))             # scalar
    return th.reshape((k, 1)), float(th_0)

# Compute u_a and b_u_a using ridge_analytic
u_a, b_u_a = ridge_analytic(B, (Z - b_v), 1)

# Solution without offsets
u_a_no_b = np.dot(np.linalg.inv(np.dot(B.T, B) + 1 * np.identity(2)), np.dot(B.T, Z))

# Feature vector for the Robot movie
v_robot = np.array([[1], [10]])

# Prediction without offset
prediction_no_offset = np.dot(u_a_no_b.T, v_robot) + 5.0

# Prediction with offset
prediction_with_offset = np.dot(u_a.T, v_robot) + b_u_a

# Prepare the result
result = [float(f"{prediction_no_offset[0][0]:.2f}"), float(f"{prediction_with_offset[0][0]:.2f}")]
print(result)  # Output the predictions


[6.06, -1.99]


<ipython-input-7-5976b3a0a38a>:18: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return th.reshape((k, 1)), float(th_0)


3) What about a brand new Robot movie (not in the existing movie data) that gets average ratings from almost everyone ($b_v^{(i)}= 3$), in the two models?

In [8]:
#Code for answering (3) here
import numpy as np

# Given data
Z = np.array([[1], [1], [5], [1], [5], [5], [1]])
b_v = np.array([[3], [3], [3], [3], [3], [5], [1]])
B = np.array([[1, 10], [1, 10], [10, 1], [1, 10], [10, 1], [5, 5], [5, 5]])

# Ridge regression function
def ridge_analytic(X, Y, lam):
    (n, k) = X.shape
    xm = np.mean(X, axis=0, keepdims=True)   # 1 x k
    ym = np.mean(Y)                          # scalar
    Z = X - xm                               # n x k
    T = Y - ym                               # n x 1
    th = np.linalg.solve(np.dot(Z.T, Z) + lam * np.identity(k), np.dot(Z.T, T))
    th_0 = (ym - np.dot(xm, th))             # scalar
    return th.reshape((k, 1)), float(th_0)

# Compute u_a and b_u_a using ridge_analytic
u_a, b_u_a = ridge_analytic(B, (Z - b_v), 1)

# Solution without offsets
u_a_no_b = np.dot(np.linalg.inv(np.dot(B.T, B) + 1 * np.identity(2)), np.dot(B.T, Z))

# Feature vector for the Robot movie
v_robot = np.array([[1], [10]])

# Prediction without offset
prediction_no_offset = np.dot(u_a_no_b.T, v_robot) + 3.0

# Prediction with offset
prediction_with_offset = np.dot(u_a.T, v_robot) + b_u_a

# Prepare the result
result = [float(f"{prediction_no_offset[0][0]:.2f}"), float(f"{prediction_with_offset[0][0]:.2f}")]
print(result)  # Output the predictions



[4.06, -1.99]


<ipython-input-8-dea0f837bdd1>:18: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return th.reshape((k, 1)), float(th_0)


##3) Implementing recommender systems

Now we'll look in detail at two implementations of matrix factorization.  One is the alternating least squares algorithm, and the other is the stochastic gradient descent algorithm.

We will assume that the input data is made up of `(a, i, r)` triples, where `a` is a user index. `i` is an item index and `r` is a rating.  Here is a small example:

In [9]:
# Data is a list of (a, i, r) triples
ratings_small = \
[(0, 0, 5), (0, 1, 3), (0, 3, 1),
 (1, 0, 4), (1, 3, 1),
 (2, 0, 1), (2, 1, 1), (2, 3, 5),
 (3, 0, 1), (3, 3, 4),
 (4, 1, 1), (4, 2, 5), (4, 3, 4)]

We will assume that predictions are made using user and item offsets, that is,
$$y = {u^{(a)}}\cdot v^{(i)} + b_u^{(a)} + b_v^{(i)}$$
In this expression $b_u$ is a vector of user offsets and $b_v$ is a vector of item offsets.

### 3.1) Alternating Least Squares (ALS)
Below is a function that provides the "outer loop" of the alternating least squares algorithm.

* Define n and m from the data.
* Initialize a list of lists `us_from_v` where `us_from_v[i]` contains the indices and ratings of users who rated item `i`.
* Similarly, initialize a list of lists `vs_from_u` where `vs_from_u[a]` contains the indices and ratings of items rated by user `a`.
* Initialize the set of parameters `x` (note that the u,v entries are set randomly while the user and item offsets are set to 0) where
    * `x[0] = u`, a list of column vectors (initialized randomly) such that `u[a]` corresponds to $u^{(a)}$ as defined above.
    * `x[1] = b_u`, a column vector (initialized with 0s) equal to $b_u$ as defined above.
    * `x[2] = v`, a list of column vectors (initialized randomly) such that `v[i]` corresponds to $v^{(i)}$ as defined above.
    * `x[3] = b_v`, a column vector (initialized with 0s) equal to $b_v$ as defined above.  
* Then we alternate minimizations.
* And report the results: the error between predicted scores and a held-out set of actual scores on the same users and items.


In [10]:
# The ALS outer loop
def mf_als(data_train, data_validate, k=2, lam=0.02, max_iter=100, verbose=False):
    # size of the problem
    n = max(d[0] for d in data_train)+1 # users
    m = max(d[1] for d in data_train)+1 # items
    # which entries are set in each row and column
    us_from_v = [[] for i in range(m)]
    vs_from_u = [[] for a in range(n)]
    for (a, i, r) in data_train:
        us_from_v[i].append((a, r))
        vs_from_u[a].append((i, r))
    # Initial guess at u, b_u, v, b_v
    # Note that u and v are lists of column vectors (columns of U, V).
    x = ([np.random.normal(1/k, size=(k,1)) for a in range(n)],
          np.zeros(n),
          [np.random.normal(1/k, size=(k,1)) for i in range(m)],
          np.zeros(m))
    # Alternation, modifies the contents of x
    start_time = time.time()
    for i in range(max_iter):
        update_U(data_train, vs_from_u, x, k, lam)
        update_V(data_train, us_from_v, x, k, lam)
        if verbose:
            print('train rmse', rmse(data_train, x), 'validate rmse', data_validate and rmse(data_validate, x))
        if data_validate == None: # code is slower, print out progress
            print("Iteration {} finished. Total Elapsed Time: {:.2f}".format(i + 1, time.time() - start_time))
    # The root mean square errors measured on validate set
    if data_validate != None:
        print('ALS result for k =', k, ': rmse train =', rmse(data_train, x), '; rmse validate =', rmse(data_validate, x))
    return x

The function `ridge_analytic(X,Y,lam)` is defined at the top of the file. Furthermore, the above code computes RMSE via the following function:


In [11]:
# Compute the root mean square error
def rmse(data, x):
    error = 0.
    for datum in data:
        error += (datum[-1] - pred(datum, x))**2
    return np.sqrt(error/len(data))

Here is an example of `vs_from_u` for the small data set given above:

```
vs_from_u = \
[[(0, 5), (1, 3), (3, 1)],
 [(0, 4), (3, 1)],
 [(0, 1), (1, 1), (3, 5)],
 [(0, 1), (3, 4)],
 [(1, 1), (2, 5), (3, 4)]]
```



Now, the only part that's missing are the `update_U` and `update_V`
procedures.  These are very similar, so we'll just do `update_U`,
where we hold the $v^{(i)}$ constant and solve for the $u^{(a)}$.  We
have seen above that each of the steps is solving a ridge regression
problem, that is, finding a set of coefficients for a linear function,
$(\theta, \theta_0)$, so as to minimize the mean sum of squared errors
on data given by $(X,Y)$ (with regularization on the magnitude of
$\theta$).

Now, write the function `update_U(data, vs_from_u, x, k, lam)`

* `data` is a list of `(a, i, r)` triples
* `vs_from_u` is a list of lists as defined above
* `x` is a list of parameters as defined above
* `k` is an integer indicating the length of the individual u and v vectors
* `lam` is the regularization parameter

The function should update the entries in `x` corresponding to the `u` vectors and the `b_u` entries.  It should also return `x`, so the Tutor can check it.
Note that if there are no ratings from a particular user, we don't want to update that user's entries.


In [12]:
def update_U(data, vs_from_u, x, k, lam):
    (u, b_u, v, b_v) = x
    # Your code here
    # Your code here
    for a in range(len(u)):
        if not vs_from_u[a]: continue
        V = np.hstack([v[i] for (i, _) in vs_from_u[a]]).T
        y = np.array([r-b_v[i] for (i, r) in vs_from_u[a]])
        u[a], b_u[a] = ridge_analytic(V, y, lam)
    return x


Here is a function to help test your code. It uses `ratings_small`, defined above.

In [13]:
def update_U_test():
  '''
  This is a test function provided to help you debug your implementation
  '''
  k = 2
  lam = 0.01

  vs_from_u = \
  [[(0, 5), (1, 3), (3, 1)],
   [(0, 4), (3, 1)],
   [(0, 1), (1, 1), (3, 5)],
   [(0, 1), (3, 4)],
   [(1, 1), (2, 5), (3, 4)]]

  np.random.seed(0)

  first = []
  for i in range(5):
    first.append(np.random.rand(2, 1))
  second = np.zeros((5,))
  third = []
  for i in range(5):
    third.append(np.random.rand(2, 1))
  fourth = np.zeros((5,))
  x0 = (first, second, third, fourth)

  x_result = update_U(ratings_small, vs_from_u, x0, k, lam)

  assert np.all(np.isclose(x_result[0], np.array([[[4.048442188078757], [-2.5000082235465526]],
                                                [[3.2715388359271054], [-1.2879317400952521]],
                                                [[-6.237522315961142], [-2.9639103597721355]],
                                                [[-3.2715388359271054], [1.2879317400952521]],
                                                [[-4.87111151185168], [-1.761023196019822]] ])))
  assert np.all(np.isclose(x_result[1].reshape(-1,),
                           np.array([[3.043665230868208], [2.048616799474877], [7.462166369240114], [2.951383200525123], [5.487071919883842]]).reshape(-1,)))
  assert np.all(np.isclose(x_result[2], np.array([[[0.7917250380826646], [0.5288949197529045]],
                                       [[0.5680445610939323], [0.925596638292661]],
                                       [[0.07103605819788694], [0.08712929970154071]],
                                       [[0.02021839744032572], [0.832619845547938]],
                                       [[0.7781567509498505], [0.8700121482468192]] ])))
  assert np.all(np.isclose(x_result[3].reshape(-1,), np.array([[0.0], [0.0], [0.0], [0.0], [0.0]]).reshape(-1,)))
  print("Test passed!")

update_U_test()

Test passed!


<ipython-input-8-dea0f837bdd1>:18: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return th.reshape((k, 1)), float(th_0)


**NOTE: ** The following **does not** need to be written: it can be filled in by clicking View Answer in the update_U problem. It will be useful in part 4.

In [14]:
def update_V(data, us_from_v, x, k, lam):
    (u, b_u, v, b_v) = x
    # Your code here
    for i in range(len(v)):
        if not us_from_v[i]: continue
        V = np.hstack([u[a] for (a, _) in us_from_v[i]]).T
        y = np.array([r-b_u[a] for (a, r) in us_from_v[i]])
        v[i], b_v[i] = ridge_analytic(V, y, lam)

    return x

### 3.2) Stochastic Gradient Descent (SGD)
Alternatively, we can use Stochastic Gradient Descent directly on the objective function
$$J(U, V) = \frac{1}{2}\sum_{(a, i) \in D} (Y_{ai} - {u^{(a)}}\cdot v^{(i)} - b_u^{(a)} - b_v^{(i)})^2 + \frac{\lambda}{2} \sum_{a = 1}^n \lVert u^{(a)} \rVert^2 + \frac{\lambda}{2} \sum_{i = 1}^m  \lVert v^{(i)} \rVert^2 $$
Note, however, that this is not strictly a sum over the data.  The regularization terms are not inside the first summation and they cannot simply be moved inside, since we would end up penalizing the parameters unevenly depending on how many ratings there were for particular items by particular users.

We can, however, define vectors of values $\lambda_u^{(a)}$ and $\lambda_v^{(i)}$ so that this is equivalent to the original objective:
$$J(U, V) = \frac{1}{2}\sum_{(a, i) \in D} \left[(Y_{ai} - {u^{(a)}}\cdot v^{(i)} - b_u^{(a)} - b_v^{(i)})^2 + \lambda_u^{(a)}\lVert u^{(a)} \rVert^2 + \lambda_v^{(i)} \lVert v^{(i)} \rVert^2\right] $$

In order to make this work out, we must have had:
$$\frac{\lambda}{2} \sum_{a = 1}^n \lVert u^{(a)} \rVert^2 = \frac{1}{2}\sum_{(a, i) \in D} \lambda_u^{(a)}\lVert u^{(a)} \rVert^2 $$
matching the terms for each $a$ separately, we must have:
$$ \frac{\lambda}{2} \lVert u^{(a)} \rVert^2 = \sum_{i : (a, i) \in D} \frac{\lambda_u^{(a)}}{2} \lVert u^{(a)} \rVert^2 = \lvert \{i : (a, i) \in D\} \rvert \frac{\lambda_u^{(a)}}{2} \lVert u^{(a)} \rVert^2 $$
Then, we must have:
$$ \lambda_u^{(a)} = \frac{\lambda}{\lvert \{i : (a, i) \in D\} \rvert}$$
Similarly,
$$ \lambda_v^{(i)} = \frac{\lambda}{\lvert \{a : (a, i) \in D\} \rvert}$$


This is now strictly a sum over the data. Note that the offset terms are not regularized.

The "outer loop" of this approach is given here:

* Define n and m from the data.
* Define vectors $\lambda_u$ and $\lambda_v$.
* Initialize the set of parameters (created and initialized as in ALS).  
* Loop picking a random data entry `(a,i,r)` and taking a step down the gradient during each iteration.
* And report the results: the error between predicted scores and a held-out set of actual scores on the same users and items.

In [15]:
# The SGD outer loop
def mf_sgd(data_train, data_validate, step_size_fn, k=2, lam=0.02, max_iter=100, verbose=False):
    # size of the problem
    ndata = len(data_train)
    n = max(d[0] for d in data_train)+1
    m = max(d[1] for d in data_train)+1
    # Distribute the lambda among the users and items
    lam_uv = lam/counts(data_train,0), lam/counts(data_train,1)
    # Initial guess at u, b_u, v, b_v (also b)
    x = ([np.random.normal(1/k, size=(k,1)) for j in range(n)],
         np.zeros(n),
         [np.random.normal(1/k, size=(k,1)) for j in range(m)],
         np.zeros(m))
    di = int(max_iter/10.)
    for i in range(max_iter):
        if i%di == 0 and verbose:
            print('i=', i, 'train rmse=', rmse(data_train, x),
                  'validate rmse', data_validate and rmse(data_validate, x))
        step = step_size_fn(i)
        j = np.random.randint(ndata)            # pick data item
        sgd_step(data_train[j], x, lam_uv, step) # modify x
    print('SGD result for k =', k, ': rmse train =', rmse(data_train, x), '; rmse validate =', rmse(data_validate, x))
    return x

The missing procedure is `sgd_step` which takes a gradient step using a
noisy estimate of the gradient based on one point.  This function will
update the relevant components of `x` corresponding to `u[a]`,
`b_u[a]`, `v[i]` and `b_v[i]`.  Your function should modify the entries in `x`
and also return `x` so that the Tutor can check it.

You will need to derive the gradient for each of the variable
components of `x`, including the offsets.  Refer back to section 1 for guidance, but
remember the offsets and remember that the offset terms are not regularized.

** WARNING: In numpy `x += y` and `x -= y` can produce different results from `x = x+y` and `x = x-y`.  For consistency in checking, please don't use `x += y` or `x -= y`!
<a href="https://stackoverflow.com/questions/31987713/numpy-array-difference-between-a-x-vs-a-a-x">Read here</a> **


In [16]:
def sgd_step(data, x, lam, step):
    (a, i, r) = data
    (u, b_u, v, b_v) = x
    (lam_u, lam_v) = lam

    # Your code here
    du_a=(u[a].T@v[i]+b_u[a]+b_v[i]-r)*v[i]+lam_u[a]*u[a]
    db_u_a=u[a].T@v[i]+b_u[a]+b_v[i]-r
    dv_i=(u[a].T@v[i]+b_u[a]+b_v[i]-r)*u[a]+lam_v[i]*v[i]
    db_v_i=u[a].T@v[i]+b_u[a]+b_v[i]-r

    u[a]=u[a]-du_a*step
    b_u[a]=b_u[a]-db_u_a*step
    v[i]=v[i]-dv_i*step
    b_v[i]=b_v[i]-db_v_i*step
    x=(u, b_u, v, b_v)

    return x

Here is a function to help test your code. It uses `ratings_small`, defined above.

In [17]:
def sgd_step_test():
  '''
  This is a test function provided to help you debug your implementation
  '''
  step = 0.025
  lam =(np.array([ 0.00333333,  0.005,  0.00333333,  0.005,  0.00333333]), np.array([ 0.0025,  0.00333333,  0.01,  0.002]))

  np.random.seed(0)

  first = []
  for i in range(5):
    first.append(np.random.rand(2, 1))
  second = np.zeros((5,))
  third = []
  for i in range(5):
    third.append(np.random.rand(2, 1))
  fourth = np.zeros((5,))
  x0 = (first, second, third, fourth)

  x_result = sgd_step(ratings_small[3], x0, lam, step)

  assert np.all(np.isclose(x_result[0], np.array([[[0.5488135039273248], [0.7151893663724195]],
                                                [[0.6667107015911342], [0.5875840438721468]],
                                                [[0.4236547993389047], [0.6458941130666561]],
                                                [[0.4375872112626925], [0.8917730007820798]],
                                                [[0.9636627605010293], [0.3834415188257777]]])))
  assert np.all(np.isclose(x_result[1].reshape(-1,), np.array([[0.0], [0.08086477989447478], [0.0], [0.0], [0.0]]).reshape(-1,)))
  assert np.all(np.isclose(x_result[2], np.array([[[0.8404178830022684], [0.5729237224816648]],
                                                [[0.5680445610939323], [0.925596638292661]],
                                                [[0.07103605819788694], [0.08712929970154071]],
                                                [[0.02021839744032572], [0.832619845547938]],
                                                [[0.7781567509498505], [0.8700121482468192]]])))
  assert np.all(np.isclose(x_result[3].reshape(-1,), np.array([[0.08086477989447478], [0.0], [0.0], [0.0], [0.0]]).reshape(-1,)))
  print("Test passed!")

sgd_step_test()


Test passed!


<ipython-input-16-b7d14a8dd690>:13: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  b_u[a]=b_u[a]-db_u_a*step
<ipython-input-16-b7d14a8dd690>:15: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  b_v[i]=b_v[i]-db_v_i*step


##4) MovieLens

Now we're going to start working with some real-world data. There's a
commonly used <a
href="https://grouplens.org/datasets/movielens/">dataset</a> of
movie ratings, called the MovieLens Dataset.

Below, we have included the following utility functions:

* `load_ratings_data_small(path_data='ratings.csv')` returns
  a list of ratings triples (a, i, r) for a subset of the data, to be
  used in parameter tuning.

* `load_ratings_data(path_data='ratings.csv')` returns the full list
  of ratings triples.

* `load_movies(path_movies='movies.csv')` returns two dictionaries.
  The first maps movie indices to title strings and the second maps
  movie indices to a list of genre strings.

In [18]:
def load_ratings_data_small(path_data='ratings.csv'):
    """
    Returns two lists of triples (a, i, r) (training, validate)
    """
    # we want to "randomly" sample but make it deterministic
    def user_hash(uid):
        return 71 * uid % 401
    def user_movie_hash(uid, iid):
        return (17 * uid + 43 * iid) % 61
    data_train = []
    data_validate = []
    with open(path_data) as f_data:
        for line in f_data:
            (uid, iid, rating, timestamp) = line.strip().split(",")
            h1 = user_hash(int(uid))
            if h1 <= 40:
                h2 = user_movie_hash(int(uid), int(iid))
                if h2 <= 12:
                    data_validate.append([int(uid), int(iid), float(rating)])
                else:
                    data_train.append([int(uid), int(iid), float(rating)])
    print('Loading from', path_data,
          'users_train', len(set(x[0] for x in data_train)),
          'items_train', len(set(x[1] for x in data_train)),
          'users_validate', len(set(x[0] for x in data_validate)),
          'items_validate', len(set(x[1] for x in data_validate)))
    return data_train, data_validate

def load_ratings_data(path_data='ratings.csv'):
    """
    Returns a list of triples (a, i, r)
    """
    data = []
    with open(path_data) as f_data:
        for line in f_data:
            (uid, iid, rating, timestamp) = line.strip().split(",")
            data.append([int(uid), int(iid), float(rating)])

    print('Loading from', path_data,
          'users', len(set(x[0] for x in data)),
          'items', len(set(x[1] for x in data)))
    return data

def load_movies(path_movies='movies.csv'):
    """
    Returns a dictionary mapping item_id to item_name and another dictionary
    mapping item_id to a list of genres
    """
    data = {}
    genreMap = {}
    with open(path_movies, encoding = "utf8") as f_data:
        for line in f_data:
            parts = line.strip().split(",")
            item_id = int(parts[0])
            item_name = ",".join(parts[1:-1]) # file is poorly formatted
            item_genres = parts[-1].split("|")
            data[item_id] = item_name
            genreMap[item_id] = item_genres
    return data, genreMap

You will need to use the above functions to answer the questions below.

**NOTE: Our checkers will succeed on 95% of submissions stemming from
correct implementations.  However, it is possible (but unlikely) for
your implementation to be correct but your answer to be rejected due
to getting an unlucky initialization.  If you are fairly sure that
your code is correct, you should try re-training a new model and
submitting answers from the new model.**

If you haven't already, complete the implementations of `update_U`,
`update_V`, and `sgd_step` (if you implemented it) above based on your solutions
to the previous questions; the definition of `update_V` can be found via View Answer in
the `update_U` problem.  We will limit ourselves to running ALS in
this problem since it requires less parameter tuning than SGD does.

###4.1) Recommendations

In the following, we will want to be able to save models locally so as to avoid running als (which takes some time) too many times. In order to do this, we include the following helper functions:


In [19]:
# After retrieving the output x from mf_als, you can use this function to save the output so
# you don't have to re-train your model
def save_model(x):
    pickle.dump(x, open("ALSmodel", "wb"))

# After training and saving your model once, you can use this function to retrieve the previous model
def load_model():
    x = pickle.load(open("ALSmodel", "rb"))
    return x

Now, compute a model with ALS and save it to disk, by executing the following code:

<b> Note: If you are running into errors when running `mf_als`, make sure that your code for `update_U` did not update a user's entries if that user had no ratings. </b>

In [22]:
data = load_ratings_data()
movies_dict, genres_dict = load_movies()
model = mf_als(data, None, k=10, lam=1, max_iter=20)
save_model(model)

Loading from ratings.csv users 13366 items 2000


<ipython-input-8-dea0f837bdd1>:18: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  return th.reshape((k, 1)), float(th_0)


Iteration 1 finished. Total Elapsed Time: 4.72
Iteration 2 finished. Total Elapsed Time: 10.19
Iteration 3 finished. Total Elapsed Time: 14.86
Iteration 4 finished. Total Elapsed Time: 20.57
Iteration 5 finished. Total Elapsed Time: 25.34
Iteration 6 finished. Total Elapsed Time: 30.78
Iteration 7 finished. Total Elapsed Time: 35.65
Iteration 8 finished. Total Elapsed Time: 40.32
Iteration 9 finished. Total Elapsed Time: 45.86
Iteration 10 finished. Total Elapsed Time: 50.71
Iteration 11 finished. Total Elapsed Time: 56.12
Iteration 12 finished. Total Elapsed Time: 60.68
Iteration 13 finished. Total Elapsed Time: 66.15
Iteration 14 finished. Total Elapsed Time: 70.91
Iteration 15 finished. Total Elapsed Time: 75.55
Iteration 16 finished. Total Elapsed Time: 80.92
Iteration 17 finished. Total Elapsed Time: 85.57
Iteration 18 finished. Total Elapsed Time: 91.33
Iteration 19 finished. Total Elapsed Time: 95.94
Iteration 20 finished. Total Elapsed Time: 101.11


This takes two minutes or so on a reasonably fast laptop. As alluded to before, <b>you only have to run `mf_als` once</b> -- henceforth, you can load the saved model via:

In [23]:
model = load_model()

Note that `load_model()` returns a tuple $(u, b_u, v, b_v)$ corresponding to the matrix factorization learned using the alternating least squares algorithm from above.

We will assume this model for the rest of the questions below.

We have introduced a "synthetic" user into the ratings file with id = 270894. Remember that we can access user data in `data`, which is a list of $(a, i, r)$ tuples


1) Write a piece of code to get relevant movie ratings in `data`, then look for the movies in `genres_dict`. Based on the movies that they've rated 5.0, what is this user's favorite genre? A list of all possible genres is given by the `genres` list below:

In [24]:
genres = ['Western', 'Comedy', 'Children', 'Crime', 'Musical',
          'Adventure', 'Drama', 'Horror', 'War', 'Documentary',
          'Romance', 'Animation', 'Film-Noir', 'Sci-Fi', 'Mystery',
          'Fantasy', 'IMAX', 'Action', 'Thriller']

# Your code to find the user's favorite genre here:
data = load_ratings_data()
movies_dict, genres_dict = load_movies()
fav={}
for (a,i,r) in data:
  if a==270894 and r==5:

    for item in genres_dict[i]:
      if item in fav:
        fav[item]+=1
      else:
        fav[item]=1
    print(fav)



Loading from ratings.csv users 13366 items 2000
{'Animation': 1, 'Children': 1, 'Comedy': 1, 'Crime': 1}
{'Animation': 2, 'Children': 2, 'Comedy': 1, 'Crime': 1, 'Adventure': 1, 'Fantasy': 1, 'Musical': 1}
{'Animation': 3, 'Children': 3, 'Comedy': 1, 'Crime': 1, 'Adventure': 1, 'Fantasy': 1, 'Musical': 2}
{'Animation': 4, 'Children': 4, 'Comedy': 2, 'Crime': 1, 'Adventure': 2, 'Fantasy': 2, 'Musical': 2, 'Sci-Fi': 1, 'IMAX': 1}
{'Animation': 5, 'Children': 5, 'Comedy': 2, 'Crime': 1, 'Adventure': 3, 'Fantasy': 2, 'Musical': 2, 'Sci-Fi': 2, 'IMAX': 1}
{'Animation': 6, 'Children': 6, 'Comedy': 2, 'Crime': 1, 'Adventure': 3, 'Fantasy': 3, 'Musical': 2, 'Sci-Fi': 2, 'IMAX': 2}
{'Animation': 7, 'Children': 7, 'Comedy': 2, 'Crime': 1, 'Adventure': 3, 'Fantasy': 3, 'Musical': 2, 'Sci-Fi': 2, 'IMAX': 2, 'Drama': 1}
{'Animation': 8, 'Children': 8, 'Comedy': 3, 'Crime': 1, 'Adventure': 4, 'Fantasy': 4, 'Musical': 3, 'Sci-Fi': 2, 'IMAX': 2, 'Drama': 1, 'Romance': 1}
{'Animation': 9, 'Children': 9

2) Write a piece of code to find the top 50 movies in `movies_dict` that are predicted for this user.  How many movies in this list match their favorite genre? Remember to remove movies the user has already seen.

In [25]:
#Your code to find how many movies match their favorite genre here
data = load_ratings_data()
movies_dict, genres_dict = load_movies()
model=load_model()
result={}
for i in movies_dict.keys():
  pred_r=pred((270894,i,1.0),model)
  result[i]=pred_r

  sort_result=sorted(result.items(), key = lambda kv:(kv[1], kv[0]),reverse=True)

  seen_movie={}
  for (a,i,r) in data:
    if a==270894:
      seen_movie[i]=r

  new_movie=[]
  for (i,r) in sort_result:
    if seen_movie.get(i)==None:
      new_movie.append(i)

  top50=new_movie[0:50]
  print(top50)
  print(genres_dict[11])

  fav={}
  for i in top50:
    if genres_dict.get(i)!=None:
      for item in genres_dict[i]:
            if item in fav:
              fav[item]+=1
            else:
              fav[item]=1

  print(fav)






Streaming output truncated to the last 5000 lines.
['Comedy', 'Drama', 'Romance']
{'Adventure': 7, 'Animation': 7, 'Children': 11, 'Comedy': 14, 'Musical': 6, 'Fantasy': 5, 'Romance': 8, 'Drama': 28, 'Crime': 10, 'Mystery': 6, 'Sci-Fi': 7, 'Thriller': 11, 'Action': 7, 'Horror': 2, 'Documentary': 1, 'War': 2}
[588, 720, 745, 551, 362, 616, 783, 107, 475, 249, 318, 32, 26, 293, 471, 314, 497, 290, 34, 215, 482, 29, 47, 593, 253, 477, 50, 480, 596, 162, 592, 589, 41, 778, 337, 260, 527, 58, 235, 82, 728, 198, 81, 678, 541, 43, 765, 410, 272, 509]
['Comedy', 'Drama', 'Romance']
{'Adventure': 7, 'Animation': 7, 'Children': 11, 'Comedy': 14, 'Musical': 6, 'Fantasy': 5, 'Romance': 8, 'Drama': 28, 'Crime': 10, 'Mystery': 6, 'Sci-Fi': 7, 'Thriller': 11, 'Action': 7, 'Horror': 2, 'Documentary': 1, 'War': 2}
[588, 720, 745, 551, 362, 616, 783, 107, 475, 249, 318, 32, 26, 293, 471, 314, 497, 290, 34, 215, 482, 29, 47, 593, 253, 477, 50, 480, 596, 162, 592, 589, 41, 778, 337, 260, 527, 58, 235, 82,

### 4.2) Similarity

Typically, movie similarities are estimated based on similarity of
genre or cast, or by human-generated recommendations.  It turns out that
training a recommender system gives us, as a by-product, a new way of
measuring similarity between movies.

First, let's think about what movie similarity might mean.  One
intuition is that two movies A and B are similar if lots of people
like both of them.  However, this criterion can be misleading.  Movies
with high average ratings will look similar since lots of people will tend to
like them.  So we will refine our criterion to be: "If users tend to
like movies A and B more than what is average for the movie and for
the user, then A and B are similar".

We can generalize the above criterion to be "If users tend to rate movies A and B similarly relative to the average, then A and B are similar". Our previous criterion
only discussed the case where ratings are high, but this new criterion should hold for low
ratings as well, for example. So in general, if movies A and B are similar, then for
every user $a$ we should expect the dot product/angle (based on the answer to the previous question)
between $u^{(a)}$ and $v^{(A)}, v^{(B)}$ to be very close. This in turn implies that the two vectors
$v^{(A)}, v^{(B)}$ are either very close together in space or very close in angle, again
depending on the answer to the previous question. This motivates us to use <a
href="https://en.wikipedia.org/wiki/Cosine_similarity">cosine
similarity</a> to compute similarities between movies, being the standard way
to measure the type of proximity we are looking for. For the rest of
this homework the similarity between movies $A$ and $B$ will be
computed as $\frac{v^{(A)} \cdot v^{(B)}}{\|v^{(A)}\| \|v^{(B)}\|}$.
**Note that the similarity can be positive or negative.**

2) Write a piece of code which identifies the movies in the dataset with
the highest similarity to a given movie, using the formula for cosine
similarity defined above. Then use it to:

a) Find the 10 movies most similar to "Star Wars: Episode IV - A New Hope (1977)" (id 260).

b) Find the 10 movies most similar to "Star Wars: Episode I - The Phantom Menace (1999)" (id 2628).

In [26]:
#Code to identify movies with highest similarity to a given movie

movies_dict, genres_dict = load_movies()
model=load_model()
(u, b_u, v, b_v)=model

va=v[2628]
similarity={}
for i in movies_dict.keys():
  similarity[i]=va.T@v[i]/(np.linalg.norm(va)*np.linalg.norm(v[i]))


sort_similarity=sorted(similarity.items(), key = lambda kv:(kv[1], kv[0]),reverse=True)
top10=sort_similarity[0:11]
print([top10[i][0] for i in range(11)])



[2628, 5378, 33493, 6934, 6365, 59615, 7318, 52722, 153, 1210, 59501]


Make sure that you also print the movie names (using `movies_dict`) so you can see that they
make sense!  These results are fairly interesting; they would be quite
clear to one who has seen the movies as the Star Wars prequels (Episodes
I-III) and the originals (Episodes IV-VI) are very different movies!
But it's interesting that our model can learn this difference.

3) Now, we look at how similar movies within the same genre are. You can
use `genres_dict` to help with this.

a) For calibration, compute the average similarity between all pairs
of movies.  **Remember not to compare a movie to itself.**

In [29]:
#Code to compute average similarity between all pairs of movies
def similarity(a,b,v):
  return v[a].T@v[b]/(np.linalg.norm(v[a])*np.linalg.norm(v[b]))

movies_dict, genres_dict = load_movies()
movie_id=list(movies_dict.keys())
n_movie=len(movie_id)
model=load_model()
(u, b_u, v, b_v)=model
print(type(movie_id))
count=0
total_sim=0
for i in range(n_movie):
  for j in range(i+1,n_movie):
    count+=1
    total_sim+=similarity(movie_id[i],movie_id[j],v)

print(total_sim/count)




<class 'list'>
[[0.00258639]]


In [30]:
  movies_dict, genres_dict = load_movies()
  movie_type={}
  model=load_model()
  (u, b_u, v, b_v)=model

  for key,value in genres_dict.items():
    for item in value:
      if item not in movie_type:
        movie_type[item]=list()
      else:
        movie_type[item].append(key)


  def similarity(a,b,v):
    return v[a].T@v[b]/(np.linalg.norm(v[a])*np.linalg.norm(v[b]))

  def avg_sim(movie_list,v):
    count=0
    total_sim=0
    n_movie=len(movie_list)
    for i in range(n_movie):
      for j in range(i+1,n_movie):
        count+=1
        total_sim+=similarity(movie_list[i],movie_list[j],v)

    return total_sim/count

  for key,value in movie_type.items():
    print(key,avg_sim(value,v))

Adventure [[0.07789952]]
Animation [[0.36543199]]
Children [[0.2865977]]
Comedy [[0.05837514]]
Fantasy [[0.08952284]]
Romance [[0.06544297]]
Drama [[0.06364122]]
Action [[0.12800325]]
Crime [[0.05583773]]
Thriller [[0.08551607]]
Horror [[0.24574669]]
Mystery [[0.06908726]]
Sci-Fi [[0.14691761]]
War [[0.11075656]]
Musical [[0.24652937]]
IMAX [[0.31288834]]
Documentary [[0.3155018]]
Western [[0.14863563]]
Film-Noir [[0.43502994]]


Now compute the average similarities across all pairs of movies within
each genre. Remember that a list of all possible genres is given by
the `genres` list defined at the top of your code file.

b) Find the genre whose movies have the highest average similarity and the value of that average similarity.

In [ ]:
#Code to find the genre with the highest average similarity and its value  - As Above
# Solution is Film-Noir [[0.43502994]] as shown above

c) Find the genre whose movies have the lowest average similarity and the value of that average similarity.

In [ ]:
#Code to find the genre with the lowest average similarity and its value
#['Crime', 0.05583773]  as shown above


These results tell us that there is enormous variation in how similar
the movies under most genres are.  But, note that the similarities are
generally *positive* for movies in a genre.

Next we look at the similarities across genres, that is, find the
average similarities across pairs of movies, where each pair has
movies from different genres.  Note that movies can belong to multiple
genres.

d) Which genre is most similar to Comedy?

In [31]:
#Code to find the genre most similar to Comedy  - Ans is ['Children', 0.04484561]
def similarity_between_gen(movie_list_1,movie_list_2,v):
    count=0
    total_sim=0
    for i in movie_list_1:
      for j in movie_list_2:
        count+=1
        total_sim+=similarity(i,j,v)

    return total_sim/count

comedy_list=movie_type['Comedy']
for key,value in movie_type.items():
  print(key,similarity_between_gen(comedy_list,value,v))


Adventure [[-0.00246376]]
Animation [[0.0303336]]
Children [[0.04484561]]
Comedy [[0.05964932]]
Fantasy [[0.0141947]]
Romance [[0.02375247]]
Drama [[-0.02758628]]
Action [[-0.02088362]]
Crime [[-0.02253166]]
Thriller [[-0.03961358]]
Horror [[-0.04946112]]
Mystery [[-0.04545759]]
Sci-Fi [[-0.03253358]]
War [[-0.06041145]]
Musical [[0.03577569]]
IMAX [[-0.0231767]]
Documentary [[0.02531137]]
Western [[-0.02094498]]
Film-Noir [[-0.05734795]]


e) Which genre is least similar to Comedy?

In [ ]:
#Code to find genre least similar to Comedy
#Ans is ['War', -0.06041145] - see above



Observe that this last similarity value is negative.